src/praisonai-code/praisonai_code/cli/features/agent_tools.py

```
@@ -0,0 +1,603 @@
+                        f"list all functions and classes in {file_path}",
+                        file_path=file_path
+                    ),
+                    timeout=10.0
+                )
+                return json.dumps(result.to_dict())
+            except asyncio.TimeoutError:
+                return json.dumps({"error": "LSP list_symbols timed out (10s)", "success": False})
+        
+        return run_sync(_list())
+    
+    def lsp_find_definition(symbol: str, file_path: str = None) -> str:
```

Ruff (0.15.21)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)